examples/tts/tts-csm.cpp

```diff
@@ -0,0 +1,297 @@
+
+    int64_t t_gb_start = ggml_time_ms(); // global start time
+    int64_t t_bb       = 0; // backbone time
+    int64_t n_bb_gen   = 0; // backbone generation count
+    int64_t t_dc       = 0; // decoder time
+    int64_t n_dc_gen   = 0; // decoder generation count
+
+    bool is_stop = false;
+    std::vector<int> generated_codes;
+
+    // backbone generation loop
+    for (int k = 0; k < params.n_predict; ++k) {
+        bool is_prompt_processing = k == 0;
+
+        if (!is_prompt_processing) {
+            // generate the next RVQ semantic token
+            batch_past_embd.n_tokens     = 1;
+            batch_past_embd.pos[0]       = n_past_bb++;
+            batch_past_embd.seq_id[0][0] = 0;
+            batch_past_embd.n_seq_id[0]  = 1;
+            batch_past_embd.logits[0]    = true;
+            std::memcpy(batch_past_embd.embd, inp_past_embd.data(), inp_past_embd.size() * sizeof(float));
+        }
+
+        int64_t t_bb_start = ggml_time_ms();
+        if (llama_decode(ctx_bb, is_prompt_processing ? batch_prompt : batch_past_embd) != 0) {
+            LOG_ERR("%s: backbone llama_decode() failed\n", __func__);
+            return 1;
+        }
+        n_bb_gen++;
+        t_bb += ggml_time_ms() - t_bb_start;
+
+        auto vocab_dc = llama_model_get_vocab(model_dc);
+        auto logits   = llama_get_logits_ith(ctx_bb, is_prompt_processing ? (batch_prompt.n_tokens - 1) : 0);
+        // for (size_t i = 0; i < 10; ++i) {
+        //     printf("%4.2f, ", logits[i]);
+        // }
+        // printf("\n");
+
+        llama_token semantic_tok = sample_greedy(logits, llama_vocab_n_tokens(vocab_dc));
+        printf("Sem token %5d : %d,", 1+(int)generated_codes.size()/32, semantic_tok);
+        generated_codes.push_back(semantic_tok);
+
+        // for (size_t i = 0; i < 10; ++i) {
+        //     printf("%4.2f, ", embd[i]);
+        // }
+        // printf("\n");
+
+
+        // decoder generation loop
+        inp_past_embd = std::vector<float>(inp_past_embd.size(), 0.0f);
+        {
+            llama_kv_self_clear(ctx_dc);
+            llama_batch batch_embd  = llama_batch_init(1, embd.size(), 1);
+            llama_batch batch_token = llama_batch_init(1, 0, 1);
+
+            // first "token" is the latent embeddings from backbone
+            {
+                batch_embd.n_tokens     = 1;
+                batch_embd.pos[0]       = 0;
+                batch_embd.seq_id[0][0] = 0;
+                batch_embd.n_seq_id[0]  = 1;
+                batch_embd.logits[0]    = false;
+                std::memcpy(batch_embd.embd, embd.data(), embd.size() * sizeof(float));
+            }
+            if (llama_decode(ctx_dc, batch_embd) != 0) {
+                LOG_ERR("%s: decoder llama_decode(embd) failed\n", __func__);
+                return 1;
+            }
+
+            // then, decode the semantic_tok to generate acoustic tokens
+            llama_token tok = semantic_tok;
+            int n_codes = 32;
```

Comment:
Based on the description in the PR, shouldn't this be:

```suggestion
            int n_codes = 31;
```

Reply:
Yeah in fact it's a bit tricky here, I should document this a bit more clear:

After the decode, we also want to get the embeddings of the 31 acoustic tokens generated, so that we can "squash" them back to one-hot vector. We could do that by:
1. Having the `audio_embd` codebook in backbone and look it up --> possible but will be a bit messy, because we need to modify the cgraph quite a lot
2. Having the `audio_embd` codebook inside the user-space code --> may requires some hacking
3. Reuse the `audio_embd` from decoder --> seems to be the simplest way

So what I ended up doing is: after the 31st acoustic token generated, I do another `decoder.decode` pass just to get lookup the `audio_embd` of that 31st token. The output logits of this pass will be discarded (and also that's why in conversion script, we have a zero codebook page padding just for this pass - so that `build_lora_mm_id` doesn't read out-of-bound)

The equivalent python is here (note: their `_embed_tokens` can do both text token and audio token embedding lookup): https://github.com/SesameAILabs/csm/blob/2d720827843b653c4d67bb4445b1c0a4f59e646f/models.py#L155-L158